numpy/core/src/multiarray/multiarraymodule.c

```diff
@@ -3807,6 +3812,7 @@ _vec_string(PyObject *NPY_UNUSED(dummy), PyObject *args, PyObject *kwds)
         result = _vec_string_with_args(char_array, type, method, args_seq);
     }
     else {
+        Py_XDECREF(type);
         PyErr_SetString(PyExc_TypeError,
                 "'args' must be a sequence of arguments");
         goto err;
```

Comment:
I cannot leave a comment there, but it would be easier to decref `type` just below in the `result == NULL` path rather than in all the error modes of `_vec_string_no_args` / `_vec_string_with_args`

Reply:
Indeed, "funcs that steal references" means they decref upon failure. Sorry for the noise.